importer/importer.go

```diff
@@ -0,0 +1,88 @@
+package importer
+
+import (
+	"github.com/goat-project/goat-proto-go"
+	"io"
+)
+
+// AccountingServiceImpl implements goat_grpc.AccountingService server
+type AccountingServiceImpl struct {
+	vmConsumer      chan<- *goat_grpc.VmRecord
+	ipConsumer      chan<- *goat_grpc.IpRecord
+	storageConsumer chan<- *goat_grpc.StorageRecord
+}
+
+// NewAccountingServiceImpl creates a grpc server that sends received data to given channels and uses clientIdentifierValidator to validate client identifiers
+func NewAccountingServiceImpl(vms chan<- *goat_grpc.VmRecord, ips chan<- *goat_grpc.IpRecord, storages chan<- *goat_grpc.StorageRecord) AccountingServiceImpl {
+	return AccountingServiceImpl{
+		vmConsumer:      vms,
+		ipConsumer:      ips,
+		storageConsumer: storages,
+	}
+}
+
+// ProcessVms is a GRPC call -- do not use!
+func (asi *AccountingServiceImpl) ProcessVms(stream goat_grpc.AccountingService_ProcessVmsServer) error {
+	for {
+		data, err := stream.Recv()
+
+		if err == io.EOF {
+			return stream.SendAndClose(&goat_grpc.Confirmation{
+				Accepted: true,
+				Msg:      "ok",
+			})
+		}
+
+		if err != nil {
+			return err
+		}
+
+		for _, vm := range data.Vms {
+			asi.vmConsumer <- vm
+		}
+	}
+}
+
+// ProcessIps is a GRPC call -- do not use!
+func (asi *AccountingServiceImpl) ProcessIps(stream goat_grpc.AccountingService_ProcessIpsServer) error {
+	for {
+		data, err := stream.Recv()
+
+		if err == io.EOF {
+			return stream.SendAndClose(&goat_grpc.Confirmation{
+				Accepted: true,
+				Msg:      "ok",
+			})
+		}
+
+		if err != nil {
+			return err
+		}
+
+		for _, ip := range data.Ips {
+			asi.ipConsumer <- ip
+		}
+	}
+}
+
+// ProcessStorage is a GRPC call -- do not use!
+func (asi *AccountingServiceImpl) ProcessStorage(stream goat_grpc.AccountingService_ProcessStorageServer) error {
```

Comment:
Since all these `Process*` methods are pretty much the same isn't it possible to loose the code duplication?

Reply:
The common error handling code could be somewhat reduced, but it makes the code a little more difficult to understand, as it uses closures. 

So I put together a proposal on what it could look like. `ProcessIps` and `ProcessStorage` would look very similar to `ProcessVms`. In this example, all `Process*`methods would use `processDelegate` for error handling and that infinite loop. I think this looses the code duplication, what do you think?

```go
func (asi *AccountingServiceImpl) ProcessVms(stream goat_grpc.AccountingService_ProcessVmsServer) error {
        asi.processDelegate(func () error {
                data, err := stream.Recv()
                if (err != nil) {
                        return err
                }
                for vm := range data.Vms {
                        asi.vmConsumer <- vm
                }
        })
}

func (asi *AccountingServiceImpl) processDelegate(recv func() error) {
        for {
                err := recv()
                if err == io.EOF {
                        return stream.SendAndClose(&goat_grpc.Confirmation{
                                Accepted: true,
                                Msg:      "ok",
                        })
                }
                if err != nil {
                        return err
                }
        }
}
```

Reply:
My initial idea was to solve this with interfaces but after a closer look I don't think that's possible. This is an interesting idea. I'm not really sure since Go is crazy but you have a `stream` variable mentioned in `processDelegate` method. wouldn't that be a problem?

Reply:
You're right, it's a problem. I could create an interface with `SendAndClose` method which would allow passing various types of streams into `processDelegate` just for the purpose of closing it.

Reply:
This is what the version with the interface looks like (I left out parts of source that are identical with the previous version):
```
type SendAndClose interface {
        SendAndClose(*goat_grpc.Confirmation) error;
}

func (asi *AccountingServiceImpl) ProcessVms(stream goat_grpc.AccountingService_ProcessVmsServer) error {
        return asi.processDelegate(stream, func () error {
                // ...
        })
}
func (asi *AccountingServiceImpl) processDelegate(stream SendAndClose, recv func() error) error {
                // ...
}
```

Reply:
You mean as a second parameter for `processDelegate`? So you would pass a closure that works with the stream and the stream itself? Isn't that weird? I mean it should work it just seems a bit off. But I don't have much experience with closures.

Reply:
The problem is, that the type of `stream` is different in each Process* method. That's why each method needs to supply a different `recv` closure. However, the `SendAndClose` method is exactly the same in all the stream classes. Think about it as poor man's inheritance: imagine all the stream classes have common ancestor with `SendAndClose` method. That's why I think it makes sense. However, there is already so much thought process behind it, that we might as well leave the code duplication as it is. That's why I started this discussion :smile:

Reply:
Sure, I get why does it have to be like that. Listen, will this be an issue once we unite those records like we talked about? That would make streams for all the methods the same type, wouldn't it? Which would render this whole discussion meaningless... :smile: